cmd/snap-update-ns/change.go

```diff
@@ -591,16 +591,115 @@ func neededChangesOld(currentProfile, desiredProfile *osutil.MountProfile) []*Ch
 		}
 	}
 
-	// Mount desired entries not reused.
+	// Mount desired entries not reused, ordering by the mimic directories they
+	// need created
+
+	var desiredNotReused []osutil.MountEntry
 	for i := range desired {
 		if !reuse[desired[i].Dir] {
-			changes = append(changes, &Change{Action: Mount, Entry: desired[i]})
+			desiredNotReused = append(desiredNotReused, desired[i])
+		}
+	}
+
+	affectedTargetCreationDirs := map[string][]osutil.MountEntry{}
+	for _, desiredEntry := range desiredNotReused {
+		parentTargetDir := filepath.Dir(desiredEntry.Dir)
+		affectedTargetCreationDirs[parentTargetDir] = append(affectedTargetCreationDirs[parentTargetDir], desiredEntry)
+	}
+
+	if len(affectedTargetCreationDirs) != 0 {
+		entriesForMimicDir := map[string][]osutil.MountEntry{}
```

Comment:
Or actually: unless I'm terribly mistaken, this condition is equivalent to `len(desiredEntry) != 0`, which means that the small snippet of code between the end of this `if` block and the function return (that is, where you create the `changes`) can also be moved inside this `if` block.

Therefore, I think it's better to invert the condition and take a shortcut:
```
if len(desiredNotReused) == 0 {
     return changes
}
```
and you can do it even before the `for` block above (since that is also meaningless, if `desiredNotReused` is empty).